src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs

```diff
@@ -985,11 +1001,13 @@ public Solution WithDocumentName(DocumentId documentId, string name)
         /// Creates a new solution instance with the document specified updated to be contained in
         /// the sequence of logical folders.
         /// </summary>
-        /// <param name="documentId">Document id.</param>
-        /// <param name="folders">Sequence of folders. <see langword="null"/> values are ignored.</param>
-        public Solution WithDocumentFolders(DocumentId documentId, IEnumerable<string?> folders)
+        public Solution WithDocumentFolders(DocumentId documentId, IEnumerable<string>? folders)
         {
-            var newState = _state.WithDocumentFolders(documentId, folders);
+            CheckContainsDocument(documentId);
+
+            var newState = _state.WithDocumentFolders(documentId,
+                folders.AsBoxedImmutableArrayWithNonNullItems() ?? throw new ArgumentNullException(nameof(folders)));
```

Comment:
```suggestion
                folders.AsBoxedImmutableArrayWithNonNullItems() ?? throw new ArgumentNullException(nameof(folders)));
```

Should this be throwing different exceptions (ArgumentException vs. ArgumentNullException) for the "the input is null" versus "non-null but has null entries"?

Reply:
ArgumentNullException is commonly used for items as well: e.g. https://github.com/dotnet/roslyn/blob/092d850a7c64c1b2234c776662f44994b9d213c4/src/Compilers/Core/Portable/Compilation/Compilation.cs#L1178